File: 力扣_C++/第四周/Sunday/every.cpp
#include<iostream>
using namespace std;

struct ListNode{
	int val;
	ListNode * next;
	ListNode() : val(0), next(nullptr) {}
	ListNode(int x) : val(x), next(nullptr) {}
	ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class every {
public:
	ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
		ListNode * l3 = new ListNode(0);
		ListNode * m = l3;
		int cur = 0, next = 0,num1=0,num2=0;
		while (l1 != NULL || l2 != NULL || next != 0){
			num1 = l1 == NULL ? 0 : l1->val;
			num2 = l2 == NULL ? 0 : l2->val;
			cur = (num1 + num2 + next) % 10;
			next = (num1 + num2 + next) / 10;
			ListNode *t = new ListNode(cur);
			m->next = t;
			m = t;
			if (l1) l1 = l1->next;
			if (l2)l2 = l2->next;
		}
		return l3->next;
	}
};

int main(){
	ListNode *h1 = new ListNode(9);
	ListNode *h2 = new ListNode(9);
	ListNode *h3 = new ListNode(9);
	ListNode *h4 = new ListNode(9);
	ListNode *h5 = new ListNode(9);
	ListNode *h6 = new ListNode(9);
	ListNode *h7 = new ListNode(9);
	h1->next = h2;
	h2->next = h3;
	h3->next = h4;
	h4->next = h5;
	h5->next = h6;
	h6->next = h7;

	ListNode *r1 = new ListNode(9);
	ListNode *r2 = new ListNode(9);
	ListNode *r3 = new ListNode(9);
	ListNode *r4 = new ListNode(9);
	r1->next = r2;
	r2->next = r3;
	r3->next = r4;
	every t;
	t.addTwoNumbers(h1, r1);
}

File: 力扣/Week7/src/Friday/every.java
package Friday;

class ListNode {
     int val;
     ListNode next;
    ListNode(int x) { val = x; }
 }

public class every {
    public boolean isPalindrome(ListNode head) {
    	if(head==null || head.next==null)
    		return true;
    	//双指针找到链表中点 ----h2为前一段终点
    	ListNode h1=head,h2=head;
    	while(h1.next!=null && h1.next.next!=null) {
    		h1=h1.next.next;
    		h2=h2.next;
    	}
    	
    	//将后一段反转
    	ListNode head2=new ListNode(0),cur=h2.next;
    	head2.next=cur;
    	while(cur.next!=null) {
    		ListNode next=cur.next;
    		cur.next=next.next;
    		next.next=head2.next;
    		head2.next=next;
    	}
    	
    	//再次遍历
    	head2=head2.next;
    	while(head!=null) {
    		if(head2.val!=head.val)
    			return false;
    		head=head.next;
    		head2=head2.next;
    	}
    	return false;
    	
   }
    public static void main(String[] args) {
		every t=new every();
		ListNode h1=new ListNode(0);
		ListNode h2=new ListNode(0);
		h1.next=h2;
		t.isPalindrome(h1);
	}
}
